### PR TITLE
Upgrade to Webpack 2 and other dependencies

### PR DESCRIPTION
**Pull request by li-kai**
# Changes

## Webpack 2
- Using a new, faster source map for development
- Using `babel-preset-env` for production, which turns modules transpilation off for tree-shaking
- About 10s faster after initial no-cache `yarn start`
- About 100kb smaller after dependency updates and new babel preset
- Shows compilation errors on webpage (but not warnings)
- Displays build folder file differences for production

## ESLint
Enabled a rule now that it's fixed.

## Other Dependencies
No breaking changes, although `react-virtualized-select` should be pegged to `2.1.0` for now, `3.0.0` seems to have some scrolling bug.
React-redux-logger now logs the differences between states. \o/

### Notes
Changed from `/utils` to `webpack.parts.js`, following [SurviveJS Webpack](https://survivejs.com/webpack/)'s guide structure (IMO it's cleaner too).
Tests keep failing because `Unexpected token import`. Tried many approaches to fixing this, no luck. Also AVA caches babel config, making it harder to debug. (@yangshun please help :persevere:!)

### CODE REVIEW

File: v3/package.json [hunk 1]
```diff
@@ -3,82 +3,83 @@
...
     "clean-webpack-plugin": "^0.1.16",
     "coveralls": "^2.13.1",
-    "cross-env": "^4.0.0",
-    "css-loader": "^0.28.0",
+    "cross-env": "^5.0.0",
+    "css-loader": "^0.28.1",
+    "cssnano": "^3.10.0",
     "enzyme": "^2.8.2",
-    "eslint": "^3.19.0",
+    "eslint": "^3.15.0",
```
Comment (yangshun) [inline v3/package.json]: Why downgrade?

File: v3/package.json [hunk 2]
```diff
@@ -3,82 +3,83 @@
...
     "eslint-import-resolver-webpack": "^0.8.1",
     "eslint-loader": "^1.7.1",
-    "eslint-plugin-ava": "^4.2.0",
     "eslint-plugin-flowtype": "^2.32.1",
     "eslint-plugin-import": "^2.2.0",
-    "eslint-plugin-jsx-a11y": "^4.0.0",
-    "eslint-plugin-react": "^6.10.3",
+    "eslint-plugin-jsx-a11y": "^3.0.2 || ^4.0.0",
+    "eslint-plugin-react": "^6.9.0",
+    "eslint_d": "^4.2.5",
```
Comment (yangshun) [inline v3/package.json]: Wow this is awesome. Gonna implement it at work hehe.

File: v3/package.json [hunk 3]
```diff
@@ -3,82 +3,83 @@
...
     "extract-text-webpack-plugin": "^2.1.0",
     "file-loader": "^0.11.1",
-    "flow-bin": "^0.45.0",
+    "flow-bin": "^0.46.0",
     "flow-scripts": "^0.2.0",
     "flow-status-webpack-plugin": "^0.1.6",
-    "happypack": "^2.2.1",
```
Comment (yangshun) [inline v3/package.json]: We need to create an issue to add happypack back in future. Actually I think happypack works with webpack 2 now. I'm using webpack 2 together with happypack at work.

File: v3/package.json [hunk 4]
```diff
@@ -3,82 +3,83 @@
...
+    "start": "cross-env NODE_ENV=development babel-node scripts/start",
+    "build": "cross-env NODE_ENV=production babel-node scripts/build",
     "build:dll": "cross-env NODE_ENV=development webpack --config=webpack/webpack.config.dll.js",
     "postinstall": "npm run build:dll && flow-scripts stub",
     "flow": "flow check",
     "lint": "npm run lint:src && npm run lint:test",
-    "lint:src": "eslint 'src/js/**/*.{js,jsx}'",
-    "lint:test": "eslint test",
+    "lint:src": "eslint_d 'src/js/**/*.{js,jsx}'",
+    "lint:test": "eslint_d test",
```
Comment (yangshun) [inline v3/package.json]: Seems like we're also linting the code inside `scripts` and `webpack`. Shall we add a command to lint it?

File: v3/scripts/build.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
+const chalk = require('chalk');
```
Comment (yangshun) [inline v3/scripts/build.js]: Is this file taken from CRA?

File: v3/webpack/webpack.config.common.js [hunk 6]
```diff
@@ -1,159 +1,67 @@
...
+    resolve: {
+      // Specify a few root paths when importing our own modules,
+      // so that we can use absolute paths in our imports.
+      // E.g. Importing our own module at `/js/path/to/module` will simply be:
+      // `import module from 'path/to/module;`
+      modules: [
+        parts.PATHS.scripts,
+        parts.PATHS.app,
+        parts.PATHS.styles,
+        'node_modules',
```
Comment (yangshun) [inline v3/webpack/webpack.config.common.js]: Can use `parts.PATHS.node` instead.

File: v3/scripts/build.js (hunk 5 above)
Comment (li-kai) [inline v3/scripts/build.js]: Most of it, yeah. Like, 90%.

File: v3/package.json (hunk 1 above)
Comment (li-kai) [inline v3/package.json]: Following airbnb's config script. They're using an outdated verison 🍤

**Pull request merged**